### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by fabiospampinato** [sindresorhus/electron-context-menu#78]
I've been profiling an app of mine and I've found that this library keeps the app busy considerably more than it should, the cause of that is that by using the `remote` module many times many _synchronous_ IPC calls are sent to the main process and back, and while that's occurring everything stops in the renderer process basically.

My use case is basically creating a few context menus from the renderer process.

1.  **Issue**: first of all if we don't pass the `window` option then that's retrieved from the main process, basically for each new context menu that's created a new IPC call is made.
    - **Solution**: it would be good to note on the readme that passing this option is important for performance.

2. **Issue**: there's a check [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L221) that checks if the `webContents` object actually exists. According to Electron's (v5) own typescript types that object always exists.
   - **Solution**: remove that check, unless maybe in previous versions of Electron that could have actually been undefined or the types are wrong? In any case if that object exists then we call the `create` function, which retrieves the `webContents` object _again_, that shouldn't happen I think.

3. **Issue**: the `webContents` object of the window is retrieved multiple times, but I don't think that object ever changes, I haven't found anything about that in the docs.
    - **Solution**: the `webContents` object should be cached, by window id, if it exists (which maybe it's always the case anyway?).

4. **Issue**: when a context menu is displayed it takes about 60ms on my system to do that, again mostly because of synchronous IPC calls. I think ideally that number should be closer to 16ms at maximum on an average computer, in order to make the app feel snappier.
   - **Solution**: I haven't looked into this much yet, but I guess there are probably some performance optimizations to be done here as well, maybe a big one could be that if a menu template didn't change at all we don't need to reconstruct the menu object again.

When implementing fixes for the first 3 issues I mentioned the time needed for the creation of a single context menu when down from ~12ms to ~2.5ms on my system. IMHO that's significant enough to deserve addressing, especially considering that those wasted ~10ms need to be multiplied by the number of context menus, in my case that added up to about 100ms, which is significant especially considering that those context menus are created at startup.

Should I submit a PR for addressing these issues?

**Comment by sindresorhus** [comment]
Thanks for opening this issue.

> the cause of that is that by using the remote module many times many synchronous IPC calls are sent to the main process and back, and while that's occurring everything stops in the renderer process basically.

It's really sad how slow the Electron IPC still is. You should definitely open an issue about that. The IPC is useless if we can't actually use it.

---

1. Yes, a note in the readme could be useful, although, I would much prefer to find a way to fix the performance problem. Are you calling `electronContextMenu()` multiple times? I think maybe my `.asyncRemote` idea mentioned below could solve this problem as we could retrieve the windows asynchronously.
2. I would manually ensure that's the case first. I don't trust the Electron types. They have a history of having lots of mistakes.
3. The Electron docs used to say that you should never store `.webContents` in a variable. I cannot find that statement anymore, so either it was omitted by mistake or the behavior changed. This should be double-checked.
4. 60ms for just showing the menu is definitely too much. I'm very curious what slows it down. Did you test with just an empty menu? Just to ensure it's not just `menu.popup` being slow?

---

> Should I submit a PR for addressing these issues?

:+1:

---

I've also been thinking about Electron having a `.asyncRemote` module that is like `.remote`, but instead wraps all the methods in a promise, so we can await them instead. No blocking. If you agree with this idea, you should definitely open a feature request on Electron.

**Comment by fabiospampinato** [comment]
> It's really sad how slow the Electron IPC still is. You should definitely open an issue about that. The IPC is useless if we can't actually use it.

Actually sending messages like so:

```js 
require ( 'electron' ).ipcRenderer.send ( 'foo' )
```

is pretty insignificant performance-wise, at least in Electron v5. I think the issue comes when passing large objects via IPC, maybe there's a big serialization overhead or something.

> I would much prefer to find a way to fix the performance problem

If I'm understanding the situation correctly the main bottleneck comes from passing large things via IPC, currently when using this library in the renderer process the following objects are retrieved from the main process:

1. `BrowserWindow` objects
	- [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L231), if not provided via the `window` option.
	- [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L211), when showing the menu
2. `WebContents` objects
	- [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L221), if the `window` option is provided.
	- [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L32), for listening to the `context-menu` event.
3. The "finished" `Menu` object
	- [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L202)

And I think we are effectively sending the `Menu` object back to the main process [here](https://github.com/sindresorhus/electron-context-menu/blob/295e19dbae4b3927f199031c53f7219765d2ed96/index.js#L211) when calling it's `popup` method.

I think all that could be eliminated this way:

- If the library is used from the main process there's no change necessary as far as the user is concerned.
- If the library is used from the renderer process:
	- We ask the user to run `require ('electron-context-menu' ).register ()` or something like that in the main process, which registers an IPC listener.
	- Then when the library is used from the renderer process instead of retrieving those expensive objects in the renderer process we just send a single asynchronous IPC call to the main process with the menu template and ask it to display it.

The downside is that asking the user to explicitly run the code for adding the IPC listener in the main process is not great. I guess we could run that automatically from the renderer process via something like `require ( 'electron' ).remote.require ( 'electron-context-menu/register' )`, but the user might have `remote` disabled for security, and doing so kinda defeats the purpose of minimizing IPC calls.

I don't think there's another possible solution 🤔.

> I would manually ensure that's the case first. I don't trust the Electron types

I can check if `webContents` exists before the `dom-ready` event and right after a `BrowserWindow` object is instantiated 👍 Checking if this is the case also on Electron v4 and v6 might be too much trouble though.

> The Electron docs used to say that you should never store .webContents in a variable. I cannot find that statement anymore, so either it was omitted by mistake or the behavior changed. This should be double-checked.

I think the only issue with that might be that if we keep a reference to `webContents` indefinitely in the main process then it can't get garbage collected. I'll see what I can find on this but probably just using a `WeakMap` for this should do the trick.

> 60ms for just showing the menu is definitely too much. I'm very curious what slows it down. Did you test with just an empty menu? Just to ensure it's not just menu.popup being slow?

I've taken a closer look at this, I had some interesting findings.

First of all I'm using a slightly modified version of the library that caches `webContents` objects and I'm passing the `window` option.

This is a flamechart of what happens when opening the default menu provided by this library, with no other menu enabled:

<img width="1440" alt="Screen Shot 2019-06-09 at 13 36 40" src="https://user-images.githubusercontent.com/1812093/59159110-2a0a0180-8ac5-11e9-877e-fadd9aef0ab3.png">

Notice all those synchronous IPC calls (I count 11), a bit more than I expected, suspicious.

This is a flamechart about opening the default menu, but with all 8 menus enabled (basically `shouldShowMenu` is called for every menu):

<img width="1440" alt="Screen Shot 2019-06-09 at 13 53 43" src="https://user-images.githubusercontent.com/1812093/59159120-5d4c9080-8ac5-11e9-9791-b11703f637ce.png">

Notice how every single menu has at least an IPC call associated with it, very weird, considering that we are literally just running these lines:

```js
if (typeof options.shouldShowMenu === 'function' && options.shouldShowMenu(event, props) === false) {
	return;
}
```

And my `shouldShowMenu` function is trivial.

So a bit in disbelief I made the following change:

```diff
-if (typeof options.shouldShowMenu === 'function' && options.shouldShowMenu(event, props) === false) {
+if (typeof options.shouldShowMenu === 'function' && options.shouldShowMenu(undefined, {x:0,y:0}) === false) {
```

And the IPC calls went away: 

<img width="1440" alt="Screen Shot 2019-06-09 at 13 59 40" src="https://user-images.githubusercontent.com/1812093/59159188-05faf000-8ac6-11e9-9474-2e45eccbac7e.png">

Basically even running `props.x` causes a synchronous IPC call 🙃

I'm not sure how to address this, maybe only one event handler per `webContents` object should be attached so that if I retrieve `props.x` 10 times only 1 IPC call is sent (at least I hope that's the case).

**Comment by fabiospampinato** [comment]
> I've also been thinking about Electron having a .asyncRemote module that is like .remote, but instead wraps all the methods in a promise, so we can await them instead. No blocking. If you agree with this idea, you should definitely open a feature request on Electron.

I forgot to comment on this.

It sounds interesting, but if I'm understanding Electron's architecture correctly whenever the main process is busy the renderer process is frozen basically, so even if we `await` the response on the renderer process as long as the work is done synchronously in the main process there are no benefits.

Another option might be some kind of lazy evaluation of `remote` calls, like if I'm running `require ( 'electron' ).remote.getCurrentWindow ().isVisible ()` I don't actually need the whole `BrowserWindow` object in the renderer process, I just need the main process to send me a boolean, maybe these sorts of scenarios could be optimized better.

**Comment by fabiospampinato** [comment]
Hey, sorry but I won't submit a PR for this, I've ended up rewriting the whole thing to better suit my use case, eventually I'll publish it as a standalone library too (links to the actual code below though).

I've tried sending the whole menu template to the main process, but functions can't be sent via the IPC channel because they can't be serialized properly (the closure will be lost I guess). 

The next best thing was to:

1. Attach one `context-menu` event lister per window in the main process (in order to retrieve the `params` object, otherwise we could have just listened for the `contextmenu` event from the renderer process). 

2. When the `context-menu` event is triggered the main process sends an IPC call to the renderer, passing along the `params` object (and this time retrieving it's properties won't cause any other IPC calls, I'm not entirely sure why).

3. In the renderer process once we create the first context menu we attach a single IPC event handler for catching the event we are sending from the main process.

4. Then in the renderer process as soon as we find the first menu that can be opened we stop.

This is a flamechart picturing the creation of all my context menus:

<img width="1218" alt="Screen Shot 2019-06-14 at 01 11 24" src="https://user-images.githubusercontent.com/1812093/59473990-400d1e80-8e45-11e9-9514-9028123eccd7.png">

Notice how there's not a single IPC call involved and the whole thing is about as fast at it gets.

This is a flamechart picturing what happens on right click:

<img width="958" alt="Screen Shot 2019-06-14 at 01 24 21" src="https://user-images.githubusercontent.com/1812093/59474008-587d3900-8e45-11e9-9e3e-ee9178e392e8.png">

Notice how we went from 11 IPC calls to just 2 (one for creating the menu from the template and another one for calling its `popup` method) (except the first time a menu is shown, where we retrieve `remote.Menu.buildFromTemplate` and we cache it for later use).

Files:

- [Renderer process library](https://paste.ee/p/nsJqq) and [how to use it](https://paste.ee/p/53uVk).
- [Main process library](https://paste.ee/p/4qokg) and [how to use it](https://paste.ee/p/ZcdA4).

**Comment by fabiospampinato** [comment]
@sindresorhus FYI I just stumbled upon [electron-remote](https://github.com/electron-userland/electron-remote), which kind of implements an asynchronous version of `remote`.

**Comment by Slapbox** [comment]
@sindresorhus is there any plan to update the library to rely on `ipc` now that it's been improved so much and `remote` is being removed entirely soon?

**Comment by sindresorhus** [comment]
This package now only works in the main process. It's too complicated to make it work with normal IPC.

https://github.com/sindresorhus/electron-context-menu/releases/tag/v3.0.0
